### DISCUSSION TITLE
服务端导入失败：duplicate key value violates unique constraint 及评论时间被重置为同一时间点

### DISCUSSION THREAD

**Discussion by dyxang** [walinejs/waline#1100] [Q&A] [answered]
**已知是[Postgresql问题](https://www.google.com/search?q=duplicate+key+value+violates+unique+constraint+%E8%A7%A3%E5%86%B3)**，但是因为我是小白，在尝试解决问题。

应该是Waline从Leancloud导出的"objectId"有问题，比如原本Leancloud导出的里面某个字段：`"objectId": "5e9dd39866303d00085ad345"`，实际上它只会取数字，所以导入后取了第一位数字5为id，但是后面很多评论也是`"objectId": "5XXXX"`开头，导致直接不唯一。不知道是Supbase问题还是本身postgresql问题

到这一步我就不会解决了，求帮助。

---
原先已使用LeanCloud作为Waline数据库

- 在Waline后台导入至 Postgresql 时显示`duplicate key value violates unique constraint "wl_comment_pkey"`
- 查看后台评论，只显示时间轴最早的一条评论
- 未更改过`waline.json`

大概我的json文件结构是这样

```
{
	"type": "waline",
	"version": 1,
	"time": 1654499627806,
	"tables": [
		"Comment",
		"Counter",
		"Users"
	],
	"data": {
		"Comment": [
			{
				"nick": "ABCD",
				"mail": "ABCD@a.com",
				"ua": "ABCD",
				"insertedAt": "2020-04-20T16:53:43.927Z",
				"link": "",
				"comment": "<p>ABCD</p>\n",
				"url": "/post/ABCD/",
				"isNotified": true,
				"objectId": "ABCD",
				"createdAt": "2020-04-20T16:53:44.717Z",
				"updatedAt": "2020-04-21T08:06:48.258Z"
			},
......
            {
				"nick": "ABCD",
				"mail": "ABCD@a.com",
				"ua": "ABCD",
				"insertedAt": "2020-04-21T15:52:24.353Z",
				"link": "",
				"comment": "<p>ABCD</p>\n",
				"url": "/post/ABCD/",
				"isNotified": true,
				"objectId": "ABCD",
				"createdAt": "2020-04-21T15:52:25.434Z",
				"updatedAt": "2020-04-22T08:04:50.169Z"
			}
		]
	}
}
```

**Comment by dyxang** [comment]
靠笨办法替换了LeanCloud原来数据里的 objectId 的数值为有序数列。
**虽然以后更新了大概会解决这个问题，但我存个档，可能对LeanCloud导出数据的人有帮助**
过程：

1. Leancloud后台，数据化存储——结构化数据
2. 选择Class名为`Comment`，然后新建列，列名称填`obj`，列类型`Number`，勾选自增，保存
3. 导出本地数据后，拿代码编辑器打开，用正则表达式进行文本替换`^.*(objectId).*\n`，删除包括`objectId`的所有行，然后文本替换所有`obj`为`objectId`
-----
原来的问题：
```
insertedAt timestamp(0) with time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
createdAt timestamp(0) with time zone NULL DEFAULT CURRENT_TIMESTAMP,
updatedAt timestamp(0) with time zone NULL DEFAULT CURRENT_TIMESTAMP,
```
这几个列的默认值是直接的时间，我导入的时候它把所有评论时间都重置为导入的时间了，排序和网址评论框显示全乱了……
等一个帮助

**Comment by lizheming** [comment]
晚点我看一下，应该是 id 格式适配的问题

**Comment by dyxang** [comment]
不是催，我已看到https://github.com/walinejs/waline/pull/1109
请问是解决这个问题了吗，如果解决了我就把这个问题关掉

**Comment by lizheming** [comment, marked as answer]
这个 Discussion 反馈的两个问题：
1. objectId 问题
2. 时间错乱问题
目前均已修复，请升级到 `@waline/vercel@1.18.7` 再试试
